# 04a: Single-Architecture, More-Seeds Test

B Steele, Colorado State University, ROSSyndicate — last updated Sep 16, 2026

`04_make_models.ipynb`'s production ensemble is 5 seeds x 2 algorithms (XGBoost + LightGBM) x 4 folds = 40 models. A follow-up check on the holdout predictions from that run found the two algorithms agree very closely with each other (0.998 correlation between their 5-seed ensemble predictions) while different seeds disagree noticeably more (~0.99 pairwise correlation) — a variance decomposition attributed ~65% of the ensemble's internal spread to seed and only ~13% to model architecture. That raises a real downstream-application question, worth checking early while the pipeline is still easy to change: **would one algorithm with more seeds capture as much (or more) of the useful variability as two algorithms with fewer seeds, at the same total model count?**

This notebook runs that experiment directly: **XGBoost only, 10 independent seeds** (`601`–`610`, a fresh namespace so as not to collide with `04`'s production seeds `501`–`505`), each with its own complete correlation-pruning + backward-elimination + gap-aware-tuning + training pipeline — the same per-seed pipeline as `04`, minus the joint (cross-algorithm) elimination step, which only makes sense with two algorithms. 10 seeds x 1 algorithm x 4 folds = 40 models, the same total count as the production ensemble, evaluated on the exact same fixed holdout, for a fair apples-to-apples comparison.

XGBoost is the one kept here (not LightGBM) because it was marginally the stronger of the two solo ensembles in `05_evaluate_ensemble.ipynb`'s results (RMSE 1.270 m vs. 1.274 m) — a small, not decisive, margin.

This is an exploratory side-experiment, not a replacement for `04`: its outputs live under `xg_models/04a_single_arch/`, entirely separate from `xg_models/v3_production/`.

## Setup

In [1]:
import json
import sys
import time
from pathlib import Path

import numpy as np
import pandas as pd

sys.path.insert(0, str(Path("python").resolve()))
import backward_elim as be
import features
import metrics as M
import model_xgb
import spatial_cv
from tuning import gap_aware_tune

be.register_model("xgboost", model_xgb)

AQUAMATCH_DIR = Path("aquamatch_files")
MODELS_DIR = Path("xg_models") / "04a_single_arch"
MODELS_DIR.mkdir(parents=True, exist_ok=True)

TARGET = "harmonized_value"
MODEL_NAME = "xgboost"
MOD = model_xgb
SEEDS = list(range(601, 611))  # 10 seeds, fresh namespace vs. 04's 501-505
N_TRIALS_INITIAL = 25
N_TRIALS_FINAL = 40
ELIM_TOL = 0.005
ELIM_SEED = 47
TUNE_SEED = 47

SITE_COLS = ["elevation_m", "catchment_area_sqkm", "pct_impervious_2006", "pct_urban_2006",
             "pct_forest_2006", "pct_cropland_2006", "pct_wetland_2006", "shore_flag"]
WEATHER_COLS_CURATED = ["precip_mm_prev30", "tmax_degC_prev30"]


def log(seed, msg):
    print(f"[{time.strftime('%H:%M:%S')}][seed{seed}] {msg}", flush=True)

## Load data and build 10 fresh CV-fold arrangements

Reads `03_split_data.ipynb`'s output (the same fixed holdout used everywhere else in this project) and adds 10 new `cvfold_seed{601-610}` columns via `spatial_cv.build_splits` — the holdout itself is rebuilt deterministically from the same `holdout_seed=3`, so it comes back identical to `03`'s; only the CV-pool fold arrangement is new. `04`'s original `cvfold_seed{501-505}` columns are left untouched alongside these.

`features.coarsen_site_features()` reduces `catchment_area_sqkm` (log-scale binned) and the `pct_*_2006` land-cover fractions (rounded to the nearest percentage point) so a heavily-resampled single waterbody's static, per-comid LakeCat values can't act as a de facto waterbody ID for the model to key a memorized SDD value off of. It does not (and isn't meant to) address the separate, already-documented leakage from the handful of waterbodies spanning more than one HUC8 (Lake Powell chief among them) — those stay large and distinct enough to remain identifiable after coarsening by design.

In [ ]:
df = pd.read_parquet(AQUAMATCH_DIR / "base_with_splits.parquet")
df = features.add_spectral_indices(df)
df = features.coarsen_site_features(df)
df = spatial_cv.build_splits(df, ensemble_seeds=SEEDS)
print(f"{len(df):,} rows, {df.shape[1]} columns; holdout n={df['is_holdout'].sum():,}")

## Per-seed pipeline (single algorithm)

Same stages as `04_make_models.ipynb`, minus the cross-algorithm joint elimination step (not applicable with one algorithm) — the single model's own backward-elimination result is the seed's final feature set directly, with no intersection step.

In [3]:
def run_seed_pipeline(seed, df):
    out_dir = MODELS_DIR / f"seed{seed}"
    out_dir.mkdir(parents=True, exist_ok=True)
    t_start = time.time()

    folds = spatial_cv.build_folds(df, seed)
    test = df[df["is_holdout"]].reset_index(drop=True)
    log(seed, f"folds: {[(f.part, len(f.train), len(f.val)) for f in folds]}, holdout n={len(test)}")

    # --- candidate pool + correlation pruning (this seed's own train_val pool) ---
    metadata_cols = {"HUC8", "holdout_part", "is_holdout"} | {f"cvfold_seed{s}" for s in SEEDS}
    extra_cols = set(SITE_COLS) | set(WEATHER_COLS_CURATED) | metadata_cols
    optical_feats = [f for f in features.candidate_feature_list(df) if f not in extra_cols]
    candidates = optical_feats + SITE_COLS + WEATHER_COLS_CURATED

    train_val = pd.concat([f.train for f in folds] + [folds[0].val], ignore_index=True).drop_duplicates(
        subset=["siteSR_id", "date"])
    pruned = features.correlation_prune(train_val, candidates, target=TARGET)
    log(seed, f"candidate pool: {len(candidates)} -> {len(pruned)} after correlation pruning")

    # --- stage 1: initial gap-aware tuning (fixed params for elimination) ---
    log(seed, f"=== initial gap-aware tuning on {len(pruned)}-feature pool ===")
    init_result = gap_aware_tune(MOD, folds, pruned, TARGET, n_trials=N_TRIALS_INITIAL, seed=47)
    fixed_params = init_result["best_params"]
    log(seed, f"val_rmse={init_result['best_score']:.4f} gap={init_result['best_gap']:.4f}")

    # --- stage 2: backward elimination (single algorithm, no joint step) ---
    log(seed, "=== backward elimination ===")
    elim_result = be.backward_eliminate(MODEL_NAME, folds, TARGET, fixed_params, pruned,
                                        weight_fn=None, tol=ELIM_TOL, seed=ELIM_SEED,
                                        log=lambda m: log(seed, m))
    final_feats = elim_result["final_features"]
    log(seed, f"final ({len(final_feats)}): {final_feats}")

    with open(out_dir / "backward_elim_summary.json", "w") as fh:
        json.dump(dict(starting_candidates=candidates, correlation_pruned=pruned,
                       final_features=final_feats, path=elim_result["path"]), fh, indent=2, default=str)

    # --- stage 3: final gap-aware tuning on the final feature set ---
    log(seed, f"=== final gap-aware tuning on {len(final_feats)}-feature set ===")
    tune_result = gap_aware_tune(MOD, folds, final_feats, TARGET, n_trials=N_TRIALS_FINAL, seed=TUNE_SEED)
    log(seed, f"chosen: val_rmse={tune_result['best_score']:.4f} gap={tune_result['best_gap']:.4f}")
    with open(out_dir / f"final_tune_{MODEL_NAME}.json", "w") as fh:
        json.dump({"model": MODEL_NAME, "features": final_feats, **tune_result}, fh, indent=2, default=str)

    # --- stage 4: final unweighted training + evaluation ---
    params = tune_result["best_params"]
    fold_models = MOD.train_fold_models(folds, final_feats, TARGET, params)
    for i, fm in enumerate(fold_models):
        fm.save_model(str(out_dir / f"{MODEL_NAME}_fold{i+1}.json"))

    cv_preds, cv_ys = [], []
    for f, model in zip(folds, fold_models):
        cv_preds.append(MOD.predict_ensemble([model], f.val[final_feats]))
        cv_ys.append(f.val[TARGET].values)
    cv_metrics = M.all_metrics(np.concatenate(cv_ys), np.concatenate(cv_preds))

    test_pred = MOD.predict_ensemble(fold_models, test[final_feats])
    test_metrics = M.all_metrics(test[TARGET].values, test_pred)
    log(seed, f"FINAL: cv_rmse={cv_metrics['rmse']:.4f} holdout_rmse={test_metrics['rmse']:.4f}")

    summary = {MODEL_NAME: dict(features=final_feats, params=params, cv_metrics=cv_metrics, test_metrics=test_metrics)}
    with open(out_dir / "final_eval_summary.json", "w") as fh:
        json.dump(summary, fh, indent=2, default=str)

    r = test[["siteSR_id", "date", "HUC4", "HUC8"]].copy()
    r["seed"] = seed
    r["model"] = MODEL_NAME
    r["y"] = test[TARGET].values
    r["pred"] = test_pred
    r.to_parquet(out_dir / "holdout_predictions.parquet")

    log(seed, f"SEED {seed} DONE in {time.time()-t_start:.0f}s")
    return summary

In [4]:
all_summaries = {}
for seed in SEEDS:
    all_summaries[seed] = run_seed_pipeline(seed, df)

[10:55:58][seed601] folds: [(1, 6803, 2189), (2, 6736, 2256), (3, 6718, 2274), (4, 6719, 2273)], holdout n=2218


[10:55:58][seed601] candidate pool: 45 -> 34 after correlation pruning


[10:55:58][seed601] === initial gap-aware tuning on 34-feature pool ===


[10:58:06][seed601] val_rmse=1.5009 gap=0.5226


[10:58:06][seed601] === backward elimination ===


[10:58:45][seed601]   [xgboost] step 0: n=34 cv_rmse=1.5210 shadow_floor=0.0004


[10:58:45][seed601]   [xgboost] drop 'tmin_degC_prev7' (importance -0.0046 <= shadow floor 0.0004)


[10:59:29][seed601]   [xgboost] step 1: n=33 cv_rmse=1.5168 shadow_floor=0.0005


[10:59:29][seed601]   [xgboost] drop 'precip_mm_prev3' (importance -0.0053 <= shadow floor 0.0005)


[11:00:11][seed601]   [xgboost] step 2: n=32 cv_rmse=1.5103 shadow_floor=0.0005


[11:00:11][seed601]   [xgboost] drop 'precip_mm_prev7' (importance -0.0020 <= shadow floor 0.0005)


[11:00:55][seed601]   [xgboost] step 3: n=31 cv_rmse=1.5074 shadow_floor=0.0005


[11:00:55][seed601]   [xgboost] drop 'tmax_degC_prev7' (importance -0.0028 <= shadow floor 0.0005)


[11:01:41][seed601]   [xgboost] step 4: n=30 cv_rmse=1.5052 shadow_floor=0.0006


[11:01:41][seed601]   [xgboost] drop 'swir1_corr7' (importance -0.0012 <= shadow floor 0.0006)


[11:02:26][seed601]   [xgboost] step 5: n=29 cv_rmse=1.5058 shadow_floor=0.0005


[11:02:26][seed601]   [xgboost] drop 'tmean_degC_prev7' (importance -0.0028 <= shadow floor 0.0005)


[11:03:09][seed601]   [xgboost] step 6: n=28 cv_rmse=1.5033 shadow_floor=0.0005


[11:03:09][seed601]   [xgboost] drop 'precip_mm_prev1' (importance -0.0013 <= shadow floor 0.0005)


[11:03:53][seed601]   [xgboost] step 7: n=27 cv_rmse=1.5006 shadow_floor=0.0006


[11:03:53][seed601]   [xgboost] drop 'shore_flag' (importance 0.0001 <= shadow floor 0.0006)


[11:04:25][seed601]   [xgboost] step 8: n=26 cv_rmse=1.5065 shadow_floor=0.0003


[11:04:25][seed601]   [xgboost] drop 'precip_mm_prev30' (importance -0.0002 <= shadow floor 0.0003)


[11:05:03][seed601]   [xgboost] step 9: n=25 cv_rmse=1.5028 shadow_floor=0.0005


[11:05:03][seed601]   [xgboost] stop: worst real feature 'NDWI' (0.0006) beats shadow floor (0.0005)


[11:05:03][seed601] final (27): ['atm_corr_LaSRC', 'green_corr7', 'nir_corr7', 'temp_corr7', 'srad_Wm2_prev1', 'srad_Wm2_prev7', 'tmean_degC_prev30', 'tmin_degC_prev30', 'srad_Wm2_prev30', 'BR', 'BG', 'NR', 'fai', 'NDVI', 'NDSSI', 'NDWI', 'MNDWI', 'elevation_m', 'catchment_area_sqkm', 'pct_impervious_2006', 'pct_urban_2006', 'pct_forest_2006', 'pct_cropland_2006', 'pct_wetland_2006', 'shore_flag', 'precip_mm_prev30', 'tmax_degC_prev30']


[11:05:03][seed601] === final gap-aware tuning on 27-feature set ===


[11:08:31][seed601] chosen: val_rmse=1.4849 gap=0.4885


[11:08:40][seed601] FINAL: cv_rmse=1.4956 holdout_rmse=1.2662


[11:08:40][seed601] SEED 601 DONE in 762s


[11:08:40][seed602] folds: [(1, 6734, 2258), (2, 6736, 2256), (3, 6839, 2153), (4, 6667, 2325)], holdout n=2218


[11:08:40][seed602] candidate pool: 45 -> 34 after correlation pruning


[11:08:40][seed602] === initial gap-aware tuning on 34-feature pool ===


[11:10:06][seed602] val_rmse=1.4406 gap=0.3949


[11:10:06][seed602] === backward elimination ===


[11:10:31][seed602]   [xgboost] step 0: n=34 cv_rmse=1.4505 shadow_floor=0.0002


[11:10:31][seed602]   [xgboost] drop 'precip_mm_prev7' (importance -0.0021 <= shadow floor 0.0002)


[11:10:59][seed602]   [xgboost] step 1: n=33 cv_rmse=1.4506 shadow_floor=0.0004


[11:10:59][seed602]   [xgboost] drop 'precip_mm_prev3' (importance -0.0005 <= shadow floor 0.0004)


[11:11:29][seed602]   [xgboost] step 2: n=32 cv_rmse=1.4388 shadow_floor=0.0003


[11:11:29][seed602]   [xgboost] drop 'precip_mm_prev30' (importance 0.0002 <= shadow floor 0.0003)


[11:11:56][seed602]   [xgboost] step 3: n=31 cv_rmse=1.4409 shadow_floor=0.0002


[11:11:56][seed602]   [xgboost] stop: worst real feature 'tmin_degC_prev30' (0.0003) beats shadow floor (0.0002)


[11:11:56][seed602] final (32): ['atm_corr_LaSRC', 'green_corr7', 'nir_corr7', 'swir1_corr7', 'temp_corr7', 'precip_mm_prev1', 'tmean_degC_prev7', 'srad_Wm2_prev1', 'srad_Wm2_prev7', 'tmax_degC_prev7', 'tmin_degC_prev7', 'tmean_degC_prev30', 'tmin_degC_prev30', 'srad_Wm2_prev30', 'BR', 'BG', 'NR', 'fai', 'NDVI', 'NDSSI', 'NDWI', 'MNDWI', 'elevation_m', 'catchment_area_sqkm', 'pct_impervious_2006', 'pct_urban_2006', 'pct_forest_2006', 'pct_cropland_2006', 'pct_wetland_2006', 'shore_flag', 'precip_mm_prev30', 'tmax_degC_prev30']


[11:11:56][seed602] === final gap-aware tuning on 32-feature set ===


[11:14:19][seed602] chosen: val_rmse=1.4366 gap=0.3872


[11:14:24][seed602] FINAL: cv_rmse=1.4623 holdout_rmse=1.2979


[11:14:24][seed602] SEED 602 DONE in 344s


[11:14:24][seed603] folds: [(1, 6840, 2152), (2, 6884, 2108), (3, 6378, 2614), (4, 6874, 2118)], holdout n=2218


[11:14:24][seed603] candidate pool: 45 -> 34 after correlation pruning


[11:14:24][seed603] === initial gap-aware tuning on 34-feature pool ===


[11:16:24][seed603] val_rmse=1.4513 gap=0.4579


[11:16:24][seed603] === backward elimination ===


[11:17:06][seed603]   [xgboost] step 0: n=34 cv_rmse=1.4607 shadow_floor=0.0004


[11:17:06][seed603]   [xgboost] drop 'precip_mm_prev3' (importance -0.0023 <= shadow floor 0.0004)


[11:17:45][seed603]   [xgboost] step 1: n=33 cv_rmse=1.4529 shadow_floor=0.0003


[11:17:45][seed603]   [xgboost] drop 'precip_mm_prev30' (importance -0.0014 <= shadow floor 0.0003)


[11:18:20][seed603]   [xgboost] step 2: n=32 cv_rmse=1.4581 shadow_floor=0.0004


[11:18:20][seed603]   [xgboost] drop 'precip_mm_prev7' (importance -0.0026 <= shadow floor 0.0004)


[11:18:58][seed603]   [xgboost] step 3: n=31 cv_rmse=1.4531 shadow_floor=0.0009


[11:18:58][seed603]   [xgboost] drop 'tmax_degC_prev7' (importance -0.0015 <= shadow floor 0.0009)


[11:19:30][seed603]   [xgboost] step 4: n=30 cv_rmse=1.4587 shadow_floor=0.0005


[11:19:30][seed603]   [xgboost] drop 'tmin_degC_prev7' (importance -0.0010 <= shadow floor 0.0005)


[11:20:05][seed603]   [xgboost] step 5: n=29 cv_rmse=1.4483 shadow_floor=0.0005


[11:20:05][seed603]   [xgboost] drop 'precip_mm_prev1' (importance -0.0005 <= shadow floor 0.0005)


[11:20:41][seed603]   [xgboost] step 6: n=28 cv_rmse=1.4454 shadow_floor=0.0002


[11:20:41][seed603]   [xgboost] drop 'swir1_corr7' (importance 0.0002 <= shadow floor 0.0002)


[11:21:17][seed603]   [xgboost] step 7: n=27 cv_rmse=1.4497 shadow_floor=0.0005


[11:21:17][seed603]   [xgboost] stop: worst real feature 'tmean_degC_prev7' (0.0006) beats shadow floor (0.0005)


[11:21:17][seed603] final (28): ['atm_corr_LaSRC', 'green_corr7', 'nir_corr7', 'swir1_corr7', 'temp_corr7', 'tmean_degC_prev7', 'srad_Wm2_prev1', 'srad_Wm2_prev7', 'tmean_degC_prev30', 'tmin_degC_prev30', 'srad_Wm2_prev30', 'BR', 'BG', 'NR', 'fai', 'NDVI', 'NDSSI', 'NDWI', 'MNDWI', 'elevation_m', 'catchment_area_sqkm', 'pct_impervious_2006', 'pct_urban_2006', 'pct_forest_2006', 'pct_cropland_2006', 'pct_wetland_2006', 'shore_flag', 'tmax_degC_prev30']


[11:21:17][seed603] === final gap-aware tuning on 28-feature set ===


[11:24:15][seed603] chosen: val_rmse=1.4364 gap=0.4374


[11:24:23][seed603] FINAL: cv_rmse=1.4427 holdout_rmse=1.2894


[11:24:23][seed603] SEED 603 DONE in 599s


[11:24:23][seed604] folds: [(1, 6459, 2533), (2, 6685, 2307), (3, 6900, 2092), (4, 6932, 2060)], holdout n=2218


[11:24:23][seed604] candidate pool: 45 -> 34 after correlation pruning


[11:24:23][seed604] === initial gap-aware tuning on 34-feature pool ===


[11:26:10][seed604] val_rmse=1.4770 gap=0.4609


[11:26:10][seed604] === backward elimination ===


[11:26:52][seed604]   [xgboost] step 0: n=34 cv_rmse=1.4871 shadow_floor=0.0006


[11:26:52][seed604]   [xgboost] drop 'tmin_degC_prev7' (importance -0.0032 <= shadow floor 0.0006)


[11:27:37][seed604]   [xgboost] step 1: n=33 cv_rmse=1.4840 shadow_floor=0.0005


[11:27:37][seed604]   [xgboost] drop 'tmax_degC_prev7' (importance -0.0034 <= shadow floor 0.0005)


[11:28:16][seed604]   [xgboost] step 2: n=32 cv_rmse=1.4811 shadow_floor=0.0004


[11:28:16][seed604]   [xgboost] drop 'precip_mm_prev7' (importance -0.0032 <= shadow floor 0.0004)


[11:28:54][seed604]   [xgboost] step 3: n=31 cv_rmse=1.4799 shadow_floor=0.0005


[11:28:54][seed604]   [xgboost] drop 'tmean_degC_prev7' (importance -0.0017 <= shadow floor 0.0005)


[11:29:34][seed604]   [xgboost] step 4: n=30 cv_rmse=1.4767 shadow_floor=0.0006


[11:29:34][seed604]   [xgboost] drop 'precip_mm_prev30' (importance -0.0012 <= shadow floor 0.0006)


[11:30:09][seed604]   [xgboost] step 5: n=29 cv_rmse=1.4769 shadow_floor=0.0005


[11:30:09][seed604]   [xgboost] drop 'srad_Wm2_prev7' (importance -0.0013 <= shadow floor 0.0005)


[11:30:44][seed604]   [xgboost] step 6: n=28 cv_rmse=1.4730 shadow_floor=0.0004


[11:30:44][seed604]   [xgboost] drop 'precip_mm_prev3' (importance -0.0001 <= shadow floor 0.0004)


[11:31:24][seed604]   [xgboost] step 7: n=27 cv_rmse=1.4716 shadow_floor=0.0006


[11:31:24][seed604]   [xgboost] drop 'precip_mm_prev1' (importance 0.0004 <= shadow floor 0.0006)


[11:31:57][seed604]   [xgboost] step 8: n=26 cv_rmse=1.4690 shadow_floor=0.0006


[11:31:57][seed604]   [xgboost] stop: worst real feature 'elevation_m' (0.0017) beats shadow floor (0.0006)


[11:31:57][seed604] final (26): ['atm_corr_LaSRC', 'green_corr7', 'nir_corr7', 'swir1_corr7', 'temp_corr7', 'srad_Wm2_prev1', 'tmean_degC_prev30', 'tmin_degC_prev30', 'srad_Wm2_prev30', 'BR', 'BG', 'NR', 'fai', 'NDVI', 'NDSSI', 'NDWI', 'MNDWI', 'elevation_m', 'catchment_area_sqkm', 'pct_impervious_2006', 'pct_urban_2006', 'pct_forest_2006', 'pct_cropland_2006', 'pct_wetland_2006', 'shore_flag', 'tmax_degC_prev30']


[11:31:57][seed604] === final gap-aware tuning on 26-feature set ===


[11:35:07][seed604] chosen: val_rmse=1.4595 gap=0.4787


[11:35:14][seed604] FINAL: cv_rmse=1.4681 holdout_rmse=1.2704


[11:35:14][seed604] SEED 604 DONE in 651s


[11:35:14][seed605] folds: [(1, 6758, 2234), (2, 6729, 2263), (3, 6751, 2241), (4, 6738, 2254)], holdout n=2218


[11:35:14][seed605] candidate pool: 45 -> 34 after correlation pruning


[11:35:14][seed605] === initial gap-aware tuning on 34-feature pool ===


[11:36:55][seed605] val_rmse=1.4574 gap=0.4581


[11:36:55][seed605] === backward elimination ===


[11:37:28][seed605]   [xgboost] step 0: n=34 cv_rmse=1.4641 shadow_floor=0.0003


[11:37:28][seed605]   [xgboost] drop 'precip_mm_prev7' (importance -0.0021 <= shadow floor 0.0003)


[11:38:07][seed605]   [xgboost] step 1: n=33 cv_rmse=1.4606 shadow_floor=0.0005


[11:38:07][seed605]   [xgboost] drop 'precip_mm_prev3' (importance -0.0024 <= shadow floor 0.0005)


[11:38:40][seed605]   [xgboost] step 2: n=32 cv_rmse=1.4518 shadow_floor=0.0003


[11:38:40][seed605]   [xgboost] drop 'precip_mm_prev30' (importance -0.0014 <= shadow floor 0.0003)


[11:39:18][seed605]   [xgboost] step 3: n=31 cv_rmse=1.4532 shadow_floor=0.0006


[11:39:18][seed605]   [xgboost] drop 'tmax_degC_prev7' (importance -0.0022 <= shadow floor 0.0006)


[11:39:49][seed605]   [xgboost] step 4: n=30 cv_rmse=1.4606 shadow_floor=0.0005


[11:39:49][seed605]   [xgboost] stop: cv_rmse degraded >0.5% from best (1.4518)


[11:39:49][seed605] final (32): ['atm_corr_LaSRC', 'green_corr7', 'nir_corr7', 'swir1_corr7', 'temp_corr7', 'precip_mm_prev1', 'tmean_degC_prev7', 'srad_Wm2_prev1', 'srad_Wm2_prev7', 'tmax_degC_prev7', 'tmin_degC_prev7', 'tmean_degC_prev30', 'tmin_degC_prev30', 'srad_Wm2_prev30', 'BR', 'BG', 'NR', 'fai', 'NDVI', 'NDSSI', 'NDWI', 'MNDWI', 'elevation_m', 'catchment_area_sqkm', 'pct_impervious_2006', 'pct_urban_2006', 'pct_forest_2006', 'pct_cropland_2006', 'pct_wetland_2006', 'shore_flag', 'precip_mm_prev30', 'tmax_degC_prev30']


[11:39:49][seed605] === final gap-aware tuning on 32-feature set ===


[11:42:32][seed605] chosen: val_rmse=1.4464 gap=0.4359


[11:42:39][seed605] FINAL: cv_rmse=1.4649 holdout_rmse=1.2753


[11:42:39][seed605] SEED 605 DONE in 445s


[11:42:39][seed606] folds: [(1, 6634, 2358), (2, 6691, 2301), (3, 6854, 2138), (4, 6797, 2195)], holdout n=2218


[11:42:39][seed606] candidate pool: 45 -> 34 after correlation pruning


[11:42:39][seed606] === initial gap-aware tuning on 34-feature pool ===


[11:44:13][seed606] val_rmse=1.4596 gap=0.4260


[11:44:13][seed606] === backward elimination ===


[11:44:44][seed606]   [xgboost] step 0: n=34 cv_rmse=1.4621 shadow_floor=0.0002


[11:44:44][seed606]   [xgboost] drop 'precip_mm_prev7' (importance -0.0010 <= shadow floor 0.0002)


[11:45:21][seed606]   [xgboost] step 1: n=33 cv_rmse=1.4643 shadow_floor=0.0004


[11:45:21][seed606]   [xgboost] drop 'precip_mm_prev3' (importance -0.0009 <= shadow floor 0.0004)


[11:45:50][seed606]   [xgboost] step 2: n=32 cv_rmse=1.4565 shadow_floor=0.0002


[11:45:50][seed606]   [xgboost] drop 'precip_mm_prev1' (importance -0.0006 <= shadow floor 0.0002)


[11:46:21][seed606]   [xgboost] step 3: n=31 cv_rmse=1.4563 shadow_floor=0.0004


[11:46:21][seed606]   [xgboost] drop 'tmax_degC_prev7' (importance -0.0009 <= shadow floor 0.0004)


[11:46:47][seed606]   [xgboost] step 4: n=30 cv_rmse=1.4577 shadow_floor=0.0004


[11:46:47][seed606]   [xgboost] stop: worst real feature 'srad_Wm2_prev7' (0.0005) beats shadow floor (0.0004)


[11:46:47][seed606] final (31): ['atm_corr_LaSRC', 'green_corr7', 'nir_corr7', 'swir1_corr7', 'temp_corr7', 'tmean_degC_prev7', 'srad_Wm2_prev1', 'srad_Wm2_prev7', 'tmax_degC_prev7', 'tmin_degC_prev7', 'tmean_degC_prev30', 'tmin_degC_prev30', 'srad_Wm2_prev30', 'BR', 'BG', 'NR', 'fai', 'NDVI', 'NDSSI', 'NDWI', 'MNDWI', 'elevation_m', 'catchment_area_sqkm', 'pct_impervious_2006', 'pct_urban_2006', 'pct_forest_2006', 'pct_cropland_2006', 'pct_wetland_2006', 'shore_flag', 'precip_mm_prev30', 'tmax_degC_prev30']


[11:46:47][seed606] === final gap-aware tuning on 31-feature set ===


[11:49:18][seed606] chosen: val_rmse=1.4495 gap=0.4170


[11:49:24][seed606] FINAL: cv_rmse=1.4594 holdout_rmse=1.2957


[11:49:24][seed606] SEED 606 DONE in 405s


[11:49:24][seed607] folds: [(1, 6833, 2159), (2, 6387, 2605), (3, 6872, 2120), (4, 6884, 2108)], holdout n=2218


[11:49:24][seed607] candidate pool: 45 -> 34 after correlation pruning


[11:49:24][seed607] === initial gap-aware tuning on 34-feature pool ===


[11:51:01][seed607] val_rmse=1.4156 gap=0.3937


[11:51:01][seed607] === backward elimination ===


[11:51:35][seed607]   [xgboost] step 0: n=34 cv_rmse=1.4221 shadow_floor=0.0003


[11:51:35][seed607]   [xgboost] drop 'precip_mm_prev3' (importance -0.0010 <= shadow floor 0.0003)


[11:52:08][seed607]   [xgboost] step 1: n=33 cv_rmse=1.4271 shadow_floor=0.0004


[11:52:08][seed607]   [xgboost] drop 'precip_mm_prev7' (importance -0.0020 <= shadow floor 0.0004)


[11:52:37][seed607]   [xgboost] step 2: n=32 cv_rmse=1.4219 shadow_floor=0.0002


[11:52:37][seed607]   [xgboost] drop 'srad_Wm2_prev7' (importance -0.0006 <= shadow floor 0.0002)


[11:53:06][seed607]   [xgboost] step 3: n=31 cv_rmse=1.4192 shadow_floor=0.0003


[11:53:06][seed607]   [xgboost] drop 'tmax_degC_prev7' (importance 0.0003 <= shadow floor 0.0003)


[11:53:37][seed607]   [xgboost] step 4: n=30 cv_rmse=1.4175 shadow_floor=0.0004


[11:53:37][seed607]   [xgboost] stop: worst real feature 'precip_mm_prev1' (0.0005) beats shadow floor (0.0004)


[11:53:37][seed607] final (30): ['atm_corr_LaSRC', 'green_corr7', 'nir_corr7', 'swir1_corr7', 'temp_corr7', 'precip_mm_prev1', 'tmean_degC_prev7', 'srad_Wm2_prev1', 'tmin_degC_prev7', 'tmean_degC_prev30', 'tmin_degC_prev30', 'srad_Wm2_prev30', 'BR', 'BG', 'NR', 'fai', 'NDVI', 'NDSSI', 'NDWI', 'MNDWI', 'elevation_m', 'catchment_area_sqkm', 'pct_impervious_2006', 'pct_urban_2006', 'pct_forest_2006', 'pct_cropland_2006', 'pct_wetland_2006', 'shore_flag', 'precip_mm_prev30', 'tmax_degC_prev30']


[11:53:37][seed607] === final gap-aware tuning on 30-feature set ===


[11:56:10][seed607] chosen: val_rmse=1.4129 gap=0.3828


[11:56:15][seed607] FINAL: cv_rmse=1.4842 holdout_rmse=1.2830


[11:56:15][seed607] SEED 607 DONE in 411s


[11:56:15][seed608] folds: [(1, 6641, 2351), (2, 6792, 2200), (3, 6785, 2207), (4, 6758, 2234)], holdout n=2218


[11:56:15][seed608] candidate pool: 45 -> 34 after correlation pruning


[11:56:15][seed608] === initial gap-aware tuning on 34-feature pool ===


[11:58:15][seed608] val_rmse=1.4829 gap=0.5093


[11:58:15][seed608] === backward elimination ===


[11:58:53][seed608]   [xgboost] step 0: n=34 cv_rmse=1.5004 shadow_floor=0.0006


[11:58:53][seed608]   [xgboost] drop 'precip_mm_prev3' (importance -0.0025 <= shadow floor 0.0006)


[11:59:34][seed608]   [xgboost] step 1: n=33 cv_rmse=1.4999 shadow_floor=0.0005


[11:59:34][seed608]   [xgboost] drop 'precip_mm_prev7' (importance -0.0023 <= shadow floor 0.0005)


[12:00:11][seed608]   [xgboost] step 2: n=32 cv_rmse=1.4987 shadow_floor=0.0004


[12:00:11][seed608]   [xgboost] drop 'precip_mm_prev30' (importance -0.0013 <= shadow floor 0.0004)


[12:00:50][seed608]   [xgboost] step 3: n=31 cv_rmse=1.4968 shadow_floor=0.0003


[12:00:50][seed608]   [xgboost] drop 'elevation_m' (importance -0.0012 <= shadow floor 0.0003)


[12:01:31][seed608]   [xgboost] step 4: n=30 cv_rmse=1.4940 shadow_floor=0.0006


[12:01:31][seed608]   [xgboost] drop 'tmax_degC_prev7' (importance -0.0013 <= shadow floor 0.0006)


[12:02:12][seed608]   [xgboost] step 5: n=29 cv_rmse=1.4880 shadow_floor=0.0004


[12:02:12][seed608]   [xgboost] drop 'precip_mm_prev1' (importance 0.0002 <= shadow floor 0.0004)


[12:02:47][seed608]   [xgboost] step 6: n=28 cv_rmse=1.4961 shadow_floor=0.0003


[12:02:47][seed608]   [xgboost] stop: cv_rmse degraded >0.5% from best (1.4880)


[12:02:47][seed608] final (29): ['atm_corr_LaSRC', 'green_corr7', 'nir_corr7', 'swir1_corr7', 'temp_corr7', 'precip_mm_prev1', 'tmean_degC_prev7', 'srad_Wm2_prev1', 'srad_Wm2_prev7', 'tmin_degC_prev7', 'tmean_degC_prev30', 'tmin_degC_prev30', 'srad_Wm2_prev30', 'BR', 'BG', 'NR', 'fai', 'NDVI', 'NDSSI', 'NDWI', 'MNDWI', 'catchment_area_sqkm', 'pct_impervious_2006', 'pct_urban_2006', 'pct_forest_2006', 'pct_cropland_2006', 'pct_wetland_2006', 'shore_flag', 'tmax_degC_prev30']


[12:02:47][seed608] === final gap-aware tuning on 29-feature set ===


[12:05:52][seed608] chosen: val_rmse=1.4814 gap=0.4577


[12:06:01][seed608] FINAL: cv_rmse=1.4867 holdout_rmse=1.2506


[12:06:01][seed608] SEED 608 DONE in 585s


[12:06:01][seed609] folds: [(1, 6770, 2222), (2, 6790, 2202), (3, 6781, 2211), (4, 6635, 2357)], holdout n=2218


[12:06:01][seed609] candidate pool: 45 -> 34 after correlation pruning


[12:06:01][seed609] === initial gap-aware tuning on 34-feature pool ===


[12:07:32][seed609] val_rmse=1.4309 gap=0.3801


[12:07:32][seed609] === backward elimination ===


[12:07:58][seed609]   [xgboost] step 0: n=34 cv_rmse=1.4409 shadow_floor=0.0003


[12:07:58][seed609]   [xgboost] drop 'precip_mm_prev3' (importance -0.0015 <= shadow floor 0.0003)


[12:08:34][seed609]   [xgboost] step 1: n=33 cv_rmse=1.4379 shadow_floor=0.0004


[12:08:34][seed609]   [xgboost] drop 'tmax_degC_prev7' (importance -0.0021 <= shadow floor 0.0004)


[12:09:09][seed609]   [xgboost] step 2: n=32 cv_rmse=1.4349 shadow_floor=0.0004


[12:09:09][seed609]   [xgboost] drop 'precip_mm_prev7' (importance -0.0024 <= shadow floor 0.0004)


[12:09:40][seed609]   [xgboost] step 3: n=31 cv_rmse=1.4286 shadow_floor=0.0004


[12:09:40][seed609]   [xgboost] drop 'tmin_degC_prev7' (importance -0.0012 <= shadow floor 0.0004)


[12:10:09][seed609]   [xgboost] step 4: n=30 cv_rmse=1.4321 shadow_floor=0.0003


[12:10:09][seed609]   [xgboost] drop 'pct_urban_2006' (importance -0.0004 <= shadow floor 0.0003)


[12:10:37][seed609]   [xgboost] step 5: n=29 cv_rmse=1.4256 shadow_floor=0.0004


[12:10:37][seed609]   [xgboost] drop 'tmean_degC_prev7' (importance -0.0010 <= shadow floor 0.0004)


[12:11:07][seed609]   [xgboost] step 6: n=28 cv_rmse=1.4196 shadow_floor=0.0004


[12:11:07][seed609]   [xgboost] drop 'srad_Wm2_prev7' (importance 0.0003 <= shadow floor 0.0004)


[12:11:35][seed609]   [xgboost] step 7: n=27 cv_rmse=1.4190 shadow_floor=0.0002


[12:11:35][seed609]   [xgboost] drop 'precip_mm_prev1' (importance -0.0002 <= shadow floor 0.0002)


[12:12:09][seed609]   [xgboost] step 8: n=26 cv_rmse=1.4227 shadow_floor=0.0005


[12:12:09][seed609]   [xgboost] drop 'precip_mm_prev30' (importance 0.0004 <= shadow floor 0.0005)


[12:12:37][seed609]   [xgboost] step 9: n=25 cv_rmse=1.4217 shadow_floor=0.0004


[12:12:37][seed609]   [xgboost] drop 'NDWI' (importance -0.0000 <= shadow floor 0.0004)


[12:13:03][seed609]   [xgboost] step 10: n=24 cv_rmse=1.4171 shadow_floor=0.0004


[12:13:03][seed609]   [xgboost] stop: worst real feature 'shore_flag' (0.0016) beats shadow floor (0.0004)


[12:13:03][seed609] final (24): ['atm_corr_LaSRC', 'green_corr7', 'nir_corr7', 'swir1_corr7', 'temp_corr7', 'srad_Wm2_prev1', 'tmean_degC_prev30', 'tmin_degC_prev30', 'srad_Wm2_prev30', 'BR', 'BG', 'NR', 'fai', 'NDVI', 'NDSSI', 'MNDWI', 'elevation_m', 'catchment_area_sqkm', 'pct_impervious_2006', 'pct_forest_2006', 'pct_cropland_2006', 'pct_wetland_2006', 'shore_flag', 'tmax_degC_prev30']


[12:13:03][seed609] === final gap-aware tuning on 24-feature set ===


[12:15:36][seed609] chosen: val_rmse=1.4092 gap=0.3802


[12:15:42][seed609] FINAL: cv_rmse=1.4417 holdout_rmse=1.2853


[12:15:42][seed609] SEED 609 DONE in 581s


[12:15:42][seed610] folds: [(1, 6670, 2322), (2, 6868, 2124), (3, 6782, 2210), (4, 6656, 2336)], holdout n=2218


[12:15:42][seed610] candidate pool: 45 -> 34 after correlation pruning


[12:15:42][seed610] === initial gap-aware tuning on 34-feature pool ===


[12:17:54][seed610] val_rmse=1.4440 gap=0.5003


[12:17:54][seed610] === backward elimination ===


[12:18:29][seed610]   [xgboost] step 0: n=34 cv_rmse=1.4583 shadow_floor=0.0004


[12:18:29][seed610]   [xgboost] drop 'precip_mm_prev30' (importance -0.0030 <= shadow floor 0.0004)


[12:19:01][seed610]   [xgboost] step 1: n=33 cv_rmse=1.4647 shadow_floor=0.0005


[12:19:01][seed610]   [xgboost] drop 'pct_urban_2006' (importance -0.0019 <= shadow floor 0.0005)


[12:19:42][seed610]   [xgboost] step 2: n=32 cv_rmse=1.4560 shadow_floor=0.0007


[12:19:42][seed610]   [xgboost] drop 'tmax_degC_prev7' (importance -0.0014 <= shadow floor 0.0007)


[12:20:23][seed610]   [xgboost] step 3: n=31 cv_rmse=1.4491 shadow_floor=0.0004


[12:20:23][seed610]   [xgboost] drop 'precip_mm_prev7' (importance -0.0003 <= shadow floor 0.0004)


[12:21:03][seed610]   [xgboost] step 4: n=30 cv_rmse=1.4538 shadow_floor=0.0004


[12:21:03][seed610]   [xgboost] drop 'precip_mm_prev1' (importance 0.0004 <= shadow floor 0.0004)


[12:21:39][seed610]   [xgboost] step 5: n=29 cv_rmse=1.4435 shadow_floor=0.0004


[12:21:39][seed610]   [xgboost] stop: worst real feature 'NDWI' (0.0013) beats shadow floor (0.0004)


[12:21:39][seed610] final (29): ['atm_corr_LaSRC', 'green_corr7', 'nir_corr7', 'swir1_corr7', 'temp_corr7', 'tmean_degC_prev7', 'srad_Wm2_prev1', 'precip_mm_prev3', 'srad_Wm2_prev7', 'tmin_degC_prev7', 'tmean_degC_prev30', 'tmin_degC_prev30', 'srad_Wm2_prev30', 'BR', 'BG', 'NR', 'fai', 'NDVI', 'NDSSI', 'NDWI', 'MNDWI', 'elevation_m', 'catchment_area_sqkm', 'pct_impervious_2006', 'pct_forest_2006', 'pct_cropland_2006', 'pct_wetland_2006', 'shore_flag', 'tmax_degC_prev30']


[12:21:39][seed610] === final gap-aware tuning on 29-feature set ===


[12:24:52][seed610] chosen: val_rmse=1.4314 gap=0.4555


[12:25:01][seed610] FINAL: cv_rmse=1.4422 holdout_rmse=1.2965


[12:25:01][seed610] SEED 610 DONE in 559s


## Per-seed summary

In [5]:
rows = []
for seed, summary in all_summaries.items():
    s = summary[MODEL_NAME]
    rows.append(dict(seed=seed, n_features=len(s["features"]), cv_rmse=s["cv_metrics"]["rmse"],
                     holdout_rmse=s["test_metrics"]["rmse"], holdout_bias=s["test_metrics"]["bias"]))
pd.DataFrame(rows).sort_values("seed")

,seed,n_features,cv_rmse,holdout_rmse,holdout_bias
0,601,27,1.495615,1.266171,0.417646
1,602,32,1.462250,1.297905,0.438523
2,603,28,1.442651,1.289409,0.434750
3,604,26,1.468134,1.270393,0.402106
4,605,32,1.464870,1.275306,0.425806
5,606,31,1.459375,1.295652,0.448578
6,607,30,1.484161,1.283035,0.442730
7,608,29,1.486745,1.250608,0.331453
8,609,24,1.441668,1.285279,0.444751
9,610,29,1.442246,1.296519,0.448917


## Assemble the 10-seed ensemble and compare against production

40 models here (10 seeds x 1 algorithm x 4 folds) vs. 40 models in `04`'s production ensemble (5 seeds x 2 algorithms x 4 folds) — same total model count, same fixed holdout, so the comparison isolates "more seeds, one algorithm" against "fewer seeds, two algorithms" directly.

In [6]:
rows_04a = []
for seed in SEEDS:
    rows_04a.append(pd.read_parquet(MODELS_DIR / f"seed{seed}" / "holdout_predictions.parquet"))
all_preds_04a = pd.concat(rows_04a, ignore_index=True)

ensemble_04a = all_preds_04a.groupby(["siteSR_id", "date"]).agg(y=("y", "first"), pred=("pred", "mean")).reset_index()
metrics_04a = M.all_metrics(ensemble_04a["y"], ensemble_04a["pred"])
print(f"04a (10 seeds x xgboost only, 40 models): {metrics_04a}")

# production ensemble, for direct comparison (04_make_models.ipynb's saved output)
PROD_DIR = Path("xg_models") / "v3_production"
rows_prod = []
for seed in [501, 502, 503, 504, 505]:
    rows_prod.append(pd.read_parquet(PROD_DIR / f"seed{seed}" / "holdout_predictions.parquet"))
all_preds_prod = pd.concat(rows_prod, ignore_index=True)
ensemble_prod = all_preds_prod.groupby(["siteSR_id", "date"]).agg(y=("y", "first"), pred=("pred", "mean")).reset_index()
metrics_prod = M.all_metrics(ensemble_prod["y"], ensemble_prod["pred"])
print(f"production (5 seeds x xgboost+lightgbm, 40 models): {metrics_prod}")

comparison = pd.DataFrame([
    dict(config="04a: 10 seeds x xgboost", **metrics_04a),
    dict(config="production: 5 seeds x xgboost+lightgbm", **metrics_prod),
])
comparison

04a (10 seeds x xgboost only, 40 models): {'rmse': 1.2668307470242317, 'mae': 0.974251922483962, 'mape': 0.4887013570318846, 'bias': 0.4040844786730668, 'p_bias': 11.403799282861936, 'smape': 0.29727696790565644, 'r2': 0.6150378657540239, 'n': 1851}
production (5 seeds x xgboost+lightgbm, 40 models): {'rmse': 1.2709690265766653, 'mae': 0.9786580894168342, 'mape': 0.4896502759021617, 'bias': 0.3996930215663622, 'p_bias': 11.279866545904042, 'smape': 0.2976288760585134, 'r2': 0.6100048093155789, 'n': 1851}


,config,rmse,mae,mape,bias,p_bias,smape,r2,n
0,04a: 10 seeds x xgboost,1.266831,0.974252,0.488701,0.404084,11.403799,0.297277,0.615038,1851
1,production: 5 seeds x xgboost+lightgbm,1.270969,0.978658,0.489650,0.399693,11.279867,0.297629,0.610005,1851


## Summary

The 10-seed x XGBoost-only ensemble (40 models) reaches holdout RMSE 1.267m / MAE 0.974m / R² 0.615 / bias +0.404m — matching, marginally beating, the production 5-seed x XGBoost+LightGBM ensemble (RMSE 1.271m / MAE 0.979m / R² 0.610 / bias +0.400m) at the same total model count. See `05a_evaluate_single_arch.ipynb` for the full evaluation (HUC8 breakdown, seed-vs-seed spread, SHAP) and the interpretation of what this means for the production pipeline's design going forward.